index.js

```diff
@@ -125,6 +136,463 @@
+
+  //  sigHighlights :: (Node, [[Boolean, (Any -> Node -> Node)]]) -> String
+  var sigHighlights = function(sig, specs) {
+    var upcarrots = '';
+    var highlightNumbers = '';
+    var number = 1;
+    var idx, spec, pos, start, end, len, showNumber, numberPos;
+    for (idx = 0; idx < specs.length; idx += 1) {
+      spec = specs[idx];
+      showNumber = spec[0];
+      pos = findNodePosition(sig)(spec[1]);
+      start = pos[0];
+      end = pos[1];
+      len = end - start;
+      numberPos = start + Math.floor((len - 1) / 2);
+      upcarrots += strRepeat(' ', start - upcarrots.length) +
+                   strRepeat('^', len);
+      highlightNumbers += strRepeat(' ', numberPos - highlightNumbers.length);
+      if (showNumber) {
+        highlightNumbers += String(number);
+        number += 1;
+      }
+    }
+    return upcarrots + '\n' + highlightNumbers;
+  };
+
+  //  sigGetName :: Node -> String
+  var sigGetName = function(sig) {
+    return sig.children[SIG_IDX_NAME].children[0].text;
+  };
+
+  //  findConstraint :: (TypeClass, String) -> Node -> Node
+  var findConstraint = function(typeClass, typeVarName) {
+    return function(sig) {
+      return traverseFind(function(node, parents) {
+        if (isVar(node) && node.text === typeVarName) {
+          var parent = getParent(parents);
+          if (parent.text === getConstraintText(typeClass)) {
+            return parent;
+          }
+        }
+      }, sig.children[SIG_IDX_CONSTRAINTS]);
+    };
+  };
+
+  //  findTypeVar :: (String, Node) -> Node
+  var findTypeVar = function(typeVarName, startNode) {
+    return traverseFind(function(node, parents) {
+      if (isVar(node) && node.text === typeVarName) {
+        return node;
+      }
+    }, startNode);
+  };
+
+  //  findArg :: (Integer, Node) -> Node
+  var findArg = function(index, sig) {
+    var args = nonDelimiters(sig.children[SIG_IDX_ARGS].children);
+    return args[index];
+  };
+
+  //  findArgTypeVar :: (Integer, String) -> Node -> Node
+  var findArgTypeVar = function(index, typeVarName) {
+    return function(sig) {
+      return findTypeVar(typeVarName, findArg(index, sig));
+    };
+  };
+
+  //  findVal :: Integer -> Node -> Node
+  var findVal = function(index) {
+    return function(sig) {
+      return isNaN(index) ? sig.children[SIG_IDX_RET] :
+                            findArg(index, sig);
+    };
+  };
+
+  //  findRetTypeVar :: String -> Node -> Node
+  var findRetTypeVar = function(typeVarName) {
+    return function(sig) {
+      return findTypeVar(typeVarName, sig.children[SIG_IDX_RET]);
+    };
+  };
+
+  //  pipe :: (Any -> [(Any -> Any)]) -> Any
+  var pipe = function(data, fns) {
+    var r = fns[0](data);
+    for (var idx = 1; idx < fns.length; idx += 1) {
+      r = fns[idx](r);
+    }
+    return r;
+  };
+
+  //  getParent :: [Node] -> Nullable Node
+  var getParent = function(parents) {
+    return last(parents);
+  };
+
+  //  head :: [a] -> Nullable a
```

Comment:
Strictly speaking the return type is not `Nullable a`, since the Nullable type differentiates between `null` and `undefined`. We could change the function's body to `return xs.length > 0 ? xs[0] : null;`, or change the type signature to something vague such as `head :: [a] -> a?`.